### PR TITLE
[INTEGRATION][SPARK] include InMemoryRelation dataset builder

### PR DESCRIPTION
**Pull request by pawel-big-lebowski**
Signed-off-by: Pawel Leszczynski <leszczynski.pawel@gmail.com>

### Problem

When a  Spark dataset is cached, its lineage is lost. 

Closes: [#610](https://github.com/OpenLineage/OpenLineage/issues/610)

### Solution

 * Get access to `CacheManager` by using reflection. List cached data and find the logical plan of the `InMemoryRelation`.
 * Provide `InMemoryRelationDatasetBuilder` that makes use of it. 
 * Prepare integration test to verify the behavior. 

> **Note:** All schema changes require discussion. Please [link the issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for context.

- [] Your change modifies the [core](https://github.com/OpenLineage/OpenLineage/blob/main/spec/OpenLineage.json) OpenLineage model
- [] Your change modifies one or more OpenLineage [facets](https://github.com/OpenLineage/OpenLineage/tree/main/spec/facets)

If you're contributing a new integration, please specify the scope of the integration and how/where it has been tested (e.g., Apache Spark integration supports `S3` and `GCS` filesystem operations, tested with AWS EMR).

### Checklist

- [x] You've [signed-off](https://github.com/OpenLineage/OpenLineage/blob/main/why-the-dco.md) your work
- [x] Your pull request title follows our [guidelines](https://github.com/OpenLineage/OpenLineage/blob/main/CONTRIBUTING.md#creating-pull-requests)
- [x] Your changes are accompanied by tests (_if relevant_)
- [x] Your change contains a [small diff](https://kurtisnusbaum.medium.com/stacked-diffs-keeping-phabricator-diffs-small-d9964f4dcfa6) and is self-contained
- [ ] You've updated any relevant documentation (_if relevant_)
- [ ] You've updated the [`CHANGELOG.md`](https://github.com/OpenLineage/OpenLineage/blob/main/CHANGELOG.md) with details about your change under the "Unreleased" section (_if relevant, depending on the change, this may not be necessary_)
- [ ] You've versioned the core OpenLineage model or facets according to [SchemaVer](https://docs.snowplowanalytics.com/docs/pipeline-components-and-applications/iglu/common-architecture/schemaver) (_if relevant_)

### CODE REVIEW

**Review by mobuchowski — APPROVED** [review]
Sometimes it's indeed needed to reflect.

File: integration/spark/src/test/resources/spark_scripts/spark_cached.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import os
+import time
+
+os.makedirs("/tmp/caching", exist_ok=True)
+
+from pyspark.sql import SparkSession
+
+spark = SparkSession.builder \
```
Comment (collado-mike) [inline integration/spark/src/test/resources/spark_scripts/spark_cached.py:10]: Can we also add the test cases described in https://github.com/OpenLineage/OpenLineage/issues/610#issuecomment-1133479889 ?

**Review by collado-mike — CHANGES_REQUESTED** [review]

File: integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/InMemoryRelationInputDatasetBuilder.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+              FieldUtils.getField(SharedState.class, "cacheManager", true).get(sharedState);
+      IndexedSeq<CachedData> cachedDataIndexedSeq =
+          (IndexedSeq<CachedData>)
+              FieldUtils.getField(CacheManager.class, "cachedData", true).get(cacheManager);
+
+      return ScalaConversionUtils.<CachedData>fromSeq(cachedDataIndexedSeq).stream()
+          .filter(
+              cachedData ->
+                  cachedData.cachedRepresentation().cacheBuilder().cachedName()
+                      == inMemoryRelation.cacheBuilder().cachedName())
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/InMemoryRelationInputDatasetBuilder.java]: :flushed: This is working bcz we get strings from the same `CachedRDDBuilder` instance. 
Thanks for spotting.

File: integration/spark/src/test/resources/spark_scripts/spark_cached.py (hunk 1 above)
Comment (pawel-big-lebowski) [reply integration/spark/src/test/resources/spark_scripts/spark_cached.py:10]: I included the test within `SparkReadWriteIntegTest` class and it is passing although I needed to change amount of events emmitted from `18` to `atLeast(12)`. The crucial part, that verifies inputs & outputs, is working properly. 

I also changed test name as I it does not seem to make use of `sqlite` nor `jdbc`.

**Review by collado-mike — APPROVED** [review]

File: integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/InMemoryRelationInputDatasetBuilder.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                              delegate(
+                                  context.getInputDatasetQueryPlanVisitors(),
+                                  context.getInputDatasetBuilders(),
+                                  event)))
+                      .stream()
+                      .flatMap(Collection::stream)
+                      .collect(Collectors.toList()))
+          .orElse(Collections.<OpenLineage.InputDataset>emptyList());
+    } catch (IllegalAccessException e) {
+      log.error("cannot extract logical plan", e);
```
[the remark this answers is not in the document]
Comment (pawel-big-lebowski) [reply integration/spark/src/main/spark3/java/io/openlineage/spark3/agent/lifecycle/plan/InMemoryRelationInputDatasetBuilder.java]: changed to warn and catching `RuntimeException`

**Pull request merged**